Clarify modifying workload resource mappings (#199)
servicebinding/spec: README.md

When a `ClusterWorkloadResourceMapping` is defined in the cluster matching a workload resource it **MUST** be used to map the binding that type. If no mapping is available for the type, the implementation **MUST** treat the workload resource as a `PodSpec`-able type.

> If a `ClusterWorkloadResourceMapping` resource is modified, and workload resources have had binding data injected using that mapping, then any future modifications to these workload resources **MUST** use the new `ClusterWorkloadResourceMapping` definitions.  Existing workloads **MUST NOT** be modified.

Reviewer: this is the opposite of my recollection from the last community call. Mandating that resources under management never be modified in perpetuity is asking for trouble.

IIRC the discussion was: during the next reconciliation, existing bindings are removed using the previous mapping and then reapplied using the new mapping. How the previous mapping is tracked through a change is left as an implementation detail (stashing as an annotation on the workload resource is likely the simplest solution).

Author: I was trying to specify that an operator shouldn't reproject all the service bindings for that workload type automatically when a new mapping is defined.  IIRC, this was a pitfall we were trying to avoid, and this was my attempt on making that behavior clear.  Perhaps I could have made that clearer in my wording.

One thing I'm not entirely clear on: what mechanism are we going to use to have users redo their existing bindings?  Should we let them update/patch existing bindings?  Or should we force them to remove and recreate their bindings?  I'm not inclined to either solution here; could we allow both?

Reviewer 2: I think it would be good to state the user scenario where someone would update the workload mapping.  The two cases that jump to mind are:
1. The referenced workload `kind` now has a new version, so the workload mapping CR is updated to have the v.next mappings
2. The user realizes they made a mistake when defining the mapping, and update the CR to correct it

Scenario 1 is simple **if** the workload mapping start with "v1", then adds "v2", etc, so then there's no effect on existing workload projections.  It gets tricky if the mapping started with a version wildcard "*" and then was updated to have specific mappings for versions - which then I think it should affect existing workload projections.

Scenario 2 should always clear out the old projections (which were incorrect) and perform the new projections.

So it may not be too bad to specify at the spec level what happens. At the implementation level though, it probably does imply the controller has to keep watching any workload mapping CRs that were used, so that it can update the projections if user scenario 1 or 2 happens.

> If a `ClusterWorkloadResourceMapping` resource is modified, and workload resources have had binding data injected using that mapping, then any modifications to these workload resources **MUST** use the new `ClusterWorkloadResourceMapping` definitions.  The existing mapping **MUST** be used to unbind existing resources, and the new mapping **MUST** be used to reproject these existing bindings.

Reviewer: Wordsmithing the proposal. Incorporate as much (or little) as you agree with.

```suggestion
After a `ClusterWorkloadResourceMapping` resource is modified and workload resources have bindings projected using that mapping, the mapping used to project the binding **MUST** be used to unproject that binding, and the new mapping used to reproject the binding.
```

Reviewer 3: I am finding the words `reproject` and `unproject` a bit confusing. Can we use more descriptive words?
For example, `reproject = project again` & `unproject = remove projection`.

Author: Another attempt at wordsmithing:
```suggestion
After a `ClusterWorkloadResourceMapping` resource is modified, existing bindings using the associated workload resource type **MUST** have their projections removed using the old mapping, and each binding **MUST** be reattempted using the new mapping.
```

Reviewer: I'm not a fan of "old" and "new" as it's possible to have multiple quick updates for a mapping. We want to remove a binding projection using the same mapping config that was use to create the projection, and then use the latest mapping to create the projection again. Depending on what actually changed in the mapping, this may be a no-op.

Taking another crack, I split the requirement into two independent requirements:
1. binding projections are removed with the same mapping used to originally create the mapping (also meaningful for deleting a ServiceBinding)
2. ClusterWorkloadResourceMapping logically remove and the reapply the binding projection (these are logical operations that can occur in a single k8s api patch/update)

```suggestion
When no longer needed, binding projections **MUST** be removed from a workload using the same mapping originally used to project binding. After a `ClusterWorkloadResourceMapping` resource is modified, each binding targeting the mapped workload type **MUST** be removed, and then reapplied with the latest mapping.
```

Reviewer 2: To throw another thing in the mix (sorry! 😄 ) , I think that `service binding resource is deleted` is limiting to just the case where the `ServiceBinding` CR is removed - but the removal of projection can also occur if the Provisioned Service is removed.  So maybe:

```
When a service binding projection is removed the controller **MUST** use the same mappings from the projection creation.  After a `ClusterWorkloadResourceMapping` resource is modified, each binding targeting the mapped workload type **MUST** be removed, then reattempted with the latest mapping.  If the workload mapping was deleted, then binding **MUST NOT** be reattempted.
```

Reviewer: > I think we should force unbinding on removal: having behavior determined by a resource that doesn't exist feels like surprising behavior, as future attempts at binding would fail, even though existing resources have successful bindings. 

I might be overfitting, but I've been trying to avoid overt language about when something should happen in order to be conformant with the spec. While most implementations will be k8s controllers that have a stream of events that will process a change within milliseconds, we also want to be open to other lifecycles, like a CLI that add/removes binding projections on a workload. A CLI inherently can't act until it is invoked to perform some action.

> Attempting to rebind with defaults, while also a possibility, most likely will not result in successful bindings, so we should probably disallow that behavior.

Let it fail. Report the errors. Someone can choose to fix the mapping, or not.

If we didn't allow this behavior, someone could create an unnecessary mapping that was bad, and then removed it expecting the system to heal itself (as k8s does). The system would refuse to undo the bad binding projections.

When a `ClusterWorkloadResourceMapping` is defined in the cluster matching a workload resource it **MUST** be used to map the binding that type.  If no mapping is available for the type, the implementation **MUST** treat the workload resource as a `PodSpec`-able type.

> When a service binding projection is removed the controller MUST use the same mappings from the projection creation. After a ClusterWorkloadResourceMapping resource is modified, each binding targeting the mapped workload type MUST be removed, then reattempted with the latest mapping.

Reviewer: ```suggestion
When a service binding projection is removed the controller **MUST** use the same mappings from the projection creation. After a `ClusterWorkloadResourceMapping` resource is modified, each binding targeting the mapped workload type **MUST** be removed, then reattempted with the latest mapping.
```

Author: Fixed.
